docs: Added architure documentation in a seperate branch (#18)
ZeroToRust/newsletter-rs: docs/architecture.md

# Email Newsletter Architecture

## 1. Introduction
### 1.1 Purpose 

> An email newsletter is system that is used to keep users update about the current changes and features of your company .

Reviewer: Add spaces between headings and other contents (wich can also be heading).

> This project is about building an email news letter.

Reviewer: ```suggestion
This project is about building an email newsletter.
```

Reviewer: Don't mark conversations as resolved without resolving them.

The the first part is to send blog post to user in a personalized manner to all clients. 

> > Note
> - The server should be able to interact with the front end to get the users detail and store it in a database.
>     
> - The app is supposed to get the users details and send post to the user based on the user's interest.

Reviewer: ```suggestion
> Note:
- The server should be able to interact with the front end to get the users detail and store it in a database.
- The app is supposed to get the users details and send post to the user based on the user's interest.
```

Reviewer: Don't mark conversations as resolved without resolving them.

Author: Ok

### 1.2 Key Features Of The Email NewsLetter
The server should support the following
 - Multiple client connection
 - User authentication endpoint
 - Create post endpoint

>  - Database connection

Reviewer: ```suggestion
 - Database for persistence
```

## 2. Architectural Overview
### 2.1 App Structure 
#### a. Very High Level Diagram

> This show how the user interacts with the system.

Reviewer: spaces

```mermaid
flowchart TD
    A((User)):::large --> B[System]
    B:::large --newsletter-->A((User))
    classDef large font-size:30px;
```

#### b. System Architecture
This diagram represent the interaction between the different components of the system

```mermaid
flowchart TD

    A(("FrontEnd (User)")) -->  B[App server]:::larger
    B --api--> A
    B --api--> C["Database "]:::large
    C --api--> B
    D[App]:::large --api--> B

>     B --api--> D

Reviewer: I believe you have reversed the positions of the app server and the app. The app is the combination of the server and the app server.

Reviewer: Don't mark conversations as resolved without resolving them.

Author: Everything is settled now

classDef large font-size:30px;
classDef larger font-size:50px;
```
- ##### Components
    - **App Server:** It is the server of the system.
    - **App:** It implement all the details of creating a blog post and sending it to the user via and email using the rust.
    - **Database:** It is used to store user credentials and personal info.
    - **Frontend:** The platform that allow the user to interact with the system.

>     ### 2.2 Tools to use Use

Reviewer 2: Repetition of "Use"

>     ### 2.2 Tools to use Use

Reviewer: Inconsistent headings and lack of spaces between heading and other contents.

>     - **Actix web:** A powerful and robust frame work that allows to build efficiently allow for easy scalability.

Reviewer 2: We're Using `axum` not `actix-web`

    - **Serde:** For serializing and deserializing data.
    - **Docker and Docker compose:** to run test locally .
    - **Postgres Database:** To manage client connections efficiently and create appropriate relationships.

    ### 2.3 End Point Specifications

>     http://locahost:7879/app/ : Root of the application

Reviewer: space

    http://localhost:7879/app/login : For login users.
    http://localhost:7879/app/login/users : For logged in users. 
    http://localhost:7879/app/blog : The menu for blog posts.
    http://localhost:7879/app/blog/create : To create a blog post.

>     http://localhost:7879/app/blog/list : To show all created blog posts

Reviewer: This should be update with the correct endpoints as the project advance.

Author: Ok

>     http://localhost:7879/app/blog/list : To show all created blog posts

Reviewer 2: Tabulate for better structuring

# Email Newsletter Architecture

## 1. Introduction
### 1.1 Purpose 

An email newsletter is system that is used to keep users update about the current changes and features of your company .

This project is about building an email news letter.
The the first part is to send blog post to user in a personalized manner to all clients. 
> Note
- The server should be able to interact with the front end to get the users detail and store it in a database.
    
- The app is supposed to get the users details and send post to the user based on the user's interest.

### 1.2 Key Features Of The Email NewsLetter

> The server should support the following

Reviewer: Please download `markdownlint` extension on VS Code to help you enforce markdown style and formatting.

 - Multiple client connection
 - User authentication endpoint
 - Create post endpoint
 - Database for persistence

## 2. Architectural Overview
### 2.1 App Structure 
#### a. Very High Level Diagram

This show how the user interacts with the system.
```mermaid
flowchart TD
    A((User)):::large --> B[System]
    B:::large --newsletter-->A((User))
    classDef large font-size:30px;
```

#### b. System Architecture
This diagram represent the interaction between the different components of the system

```mermaid
flowchart TD

    A(("FrontEnd (User)")) -->  B[App server]:::larger
    B --api--> A
    B --api--> C["Database "]:::large
    C --api--> B
    D[App]:::large --api--> B
    B --api--> D

classDef large font-size:30px;
classDef larger font-size:50px;
```
- ##### Components

    - **App Server:** It is the server of the system and precise the endpoints for hosting the different services.
    - **App:** It implement all the details of creating a blog post and sending it to the user via and email using the rust.
    - **Database:** It is used to store user credentials and personal info.
    - **Frontend:** The platform that allow the user to interact with the system.

>     ### 2.2 Tools to use Use

Reviewer: Inconsistent formatting.

Author: Installed markdownlint and formatted the code

classDef larger font-size:50px;
```
- ##### Components
    - **App Server:** It is the server of the system.
    - **App:** It implement all the details of creating a blog post and sending it to the user via and email using the rust.
    - **Database:** It is used to store user credentials and personal info.
    - **Frontend:** The platform that allow the user to interact with the system.

    ### 2.2 Tools to use Use
    - **Actix web:** A powerful and robust frame work that allows to build efficiently allow for easy scalability.
    - **Serde:** For serializing and deserializing data.
    - **Docker and Docker compose:** to run test locally .
    - **Postgres Database:** To manage client connections efficiently and create appropriate relationships.

    ### 2.3 End Point Specifications
    http://locahost:7879/app/ : Root of the application
    http://localhost:7879/app/login : For login users.
    http://localhost:7879/app/login/users : For logged in users. 
    http://localhost:7879/app/blog : The menu for blog posts.
    http://localhost:7879/app/blog/create : To create a blog post.
    http://localhost:7879/app/blog/list : To show all created blog posts

> 

Reviewer 3: Why ? Irregular spacing all across the doc.

> # Email Newsletter Architecture

Reviewer: ```suggestion
# Email Newsletter Server Architecture
```

## 1. Introduction

### 1.1 Purpose 

An email newsletter is system that is used to keep users update about the current changes and features of your company .

This project is about building an email newsletter.
The the first part is to send blog post to user in a personalized manner to all clients. 

> Note:

- The server should be able to interact with the front end to get the users detail and store it in a database.
- The app is supposed to get the users details and send post to the user based on the user's interest.

### 1.2 Key Features Of The Email NewsLetter

The server should support the following

- Multiple client connection
- User authentication endpoint

> - Create post endpoint

Reviewer: To do what?

Reviewer: Create a post endpoint to do what?

- Database for persistence

## 2. Architectural Overview

### 2.1 App Structure

#### a. Very High Level Diagram

This show how the user interacts with the system.

```mermaid
flowchart TD
    A((User)):::large --> B[System]
    B:::large --newsletter-->A((User))
    classDef large font-size:30px;
```

#### b. System Architecture

This diagram represent the interaction between the different components of the system

```mermaid 

flowchart TD
    A(("FrontEnd (User)")) -->  B[App]:::larger
    B --api--> A
    E--api--> C["Database "]:::large
    C --api--> E
    D[App Logic]:::large --api--> E
    E --api--> D
    E --api--> B
    B --api--> E[App Server]:::large

classDef large font-size:30px;
classDef larger font-size:50px; 
```

#### Components

- **App Server:** It is the server of the system and precise the endpoints for hosting the different services.
- **App:** It implement all the details of creating a blog post and sending it to the user via and email using the rust.
- **Database:** It is used to store user credentials and personal info.

> - **Frontend:** The platform that allow the user to interact with the system.

Reviewer: This is not part of the implementation. We are only building the server.

Author: ok

### 2.2 Tools to use Use

- **Axum:** A powerful and robust frame work that allows to build efficiently allow for easy scalability.
- **Serde:** For serializing and deserializing data.
- **Docker and Docker compose:** to run test locally .

> - **Postgres Database:** To manage client connections efficiently and create appropriate relationships.

Reviewer: Since you are listing the main dependencies used, where is `sqlx`? Which is one of the core deps as well?

Author: Ok good

### 2.3 End Point Specifications

|EndPoints|Services|
|---------|---------|
|`http://locahost:8080/app/`| Root of the application|
|`http://localhost:8080/app/login`| For login users.|
|`http://localhost:8080/app/login/users` | For logged in users. |
|`http://localhost:8080/app/blog`| The menu for blog posts.|
|`http://localhost:8080/app/blog/create` | To create a blog post.|

> |`http://localhost:8080/app/blog/list` | To show all created blog posts|

Reviewer: This is just placeholder for all handlers we will implement. Don't forget that this should be updated each time to reflect the current implementation.

Author: ok sir but that should be as we progres right?. So should we keep this pr as a draft.

Reviewer: I prefer putting it in the `TODO` column so it will be updated later with the latest changes of the project.

```mermaid 

flowchart TD
    A(("FrontEnd (User)")) -->  B[App]:::larger
    B --api--> A
    E--api--> C["Database "]:::large
    C --api--> E
    D[App Logic]:::large --api--> E
    E --api--> D
    E --api--> B
    B --api--> E[App Server]:::large

classDef large font-size:30px;
classDef larger font-size:50px; 
```

#### Components

- **App Server:** It is the server of the system and precise the endpoints for hosting the different services.
- **App Logic:** It implement all the details of creating a blog post and sending it to the user via and email using the rust.
- **Database:** It is used to store user credentials and personal info.

### 2.2 Tools and libraries to use Use

- **Axum:** A powerful and robust frame work that allows to build efficiently allow for easy scalability.
- **Tokio:** which is an asynchronous runtime for axum.
- **Tracing and chrono:** For logging everything happening on our server.
- **Serde:** For serializing and deserializing data.
- **eyre:** For well formatted and designed error display .
- **Sqlx:** As an ORM to communicate with our database directly through our app.
- **Docker and Docker compose:** to run test locally .

> - **Postgres Database:** To manage client connections efficiently and create appropriate relationships.

Reviewer: It is preferable to actually list only tools used and leave dependencies.
